drgn_tools/cpuinfo.py

```diff
@@ -0,0 +1,615 @@
+
+    :returns: a dictionary of vulnerabilities with their mitigations
+    """
+    vulns = [
+        "Meltdown",
+        "Spectre_V1",
+        "Spectre_V2",
+        "SSBD",
+        "L1TF",
+        "MDS",
+        "tsx_async_abort",
+        "itlb_multihit",
+        "SRBDS",
+        "mmio_stale_data",
+        "mmio_unknown",
+        "Retbleed",
+        "spec_store_bypass",
+        "GDS",
+        "SRSO",
+    ]
+
+    if "cpu_data" in prog:
+        cpuinfo_struct = prog["cpu_data"]
+    elif "boot_cpu_data" in prog:
+        cpuinfo_struct = prog["boot_cpu_data"]
+    else:
+        raise Exception(
+            "Failed to load CPU info: no cpuinfo struct found (tried 'cpu_data' and 'boot_cpu_data')"
+        )
+
+    cpu_caps_bugs = cpuinfo_struct.x86_capability
+
+    mitigations = dict.fromkeys(vulns, "Not Affected")
+
+    cpuinfo_data = x86_get_cpu_info(prog)
+    bugs = cpuinfo_data["BUG FLAGS"].split()
+
+    for bug in bugs:
+        if bug == "cpu_meltdown":
+            mitigations["Meltdown"] = get_meltdown_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "spectre_v1":
+            mitigations["Spectre_V1"] = get_spectre_v1_mitigation(prog)
+
+        elif bug == "spectre_v2":
+            mitigations["Spectre_V2"] = get_spectre_v2_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "spec_store_bypass":
+            mitigations["spec_store_bypass"] = get_ssb_mitigation(prog)
+
+        elif bug == "l1tf":
+            mitigations["L1TF"] = get_l1tf_mitigation(prog, cpu_caps_bugs)
+
+        elif bug == "mds":
+            mitigations["MDS"] = get_mds_mitigation(prog, cpu_caps_bugs)
+
+        elif bug == "taa":
+            mitigations["tsx_async_abort"] = get_taa_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "itlb_multihit":
+            mitigations["itlb_multihit"] = get_itlb_multihit_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "srbds":
+            mitigations["SRBDS"] = get_srbds_mitigation(prog)
+
+        elif bug == "mmio_stale_data":
+            mitigations["mmio_stale_data"] = get_mmio_stale_data_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "mmio_unknown":
+            mitigations["mmio_unknown"] = get_mmio_stale_data_mitigation(
+                prog, cpu_caps_bugs
+            )
+
+        elif bug == "retbleed":
+            mitigations["Retbleed"] = get_retbleed_mitigation(
+                prog, cpuinfo_struct
+            )
+
+        elif bug == "gds":
+            mitigations["GDS"] = get_gds_mitigation(prog)
+
+        elif bug == "srso":
+            mitigations["SRSO"] = "Vulnerable (Status Unknown)"
+            # Currently cannot handle this mitigation data because the kernel
+            # uses assembly code to fetch it
+
+        else:
+            mitigations[bug] = "Vulnerable"
```

Comment:
Assuming you made the hash table of mitigations, then you can just do the following:

```suggestion
    mitigations = {vuln: "Not Affected" for vuln in vuln_to_checker}
    for bug in bugs:
        vuln_name = bug_to_vuln_name.get(bug, bug)
        mitigation_checker = vuln_to_checker.get(vuln_name)
        if mitigation_checker:
            mitigations[vuln_name] = mitigation_checker(prog, cpu_caps_bugs)
        else:
            mitigations[vuln_name] = "Vulnerable"
```